### ISSUE TITLE
Activity chart days of the week are incorrect (timezone bug)

### ISSUE THREAD

**Issue by kellpossible** [muety/wakapi#719]
**Describe the bug**

The days of the week are incorrect on the activity chart tooltips, there are two mondays.

![image](https://github.com/user-attachments/assets/209b0443-d08b-46fa-8869-7998983f2528)
![image](https://github.com/user-attachments/assets/54cd600b-5424-481d-9863-69119762d1c2)

**System information**
Please provide information on:
* Wakapi version v2.12.2 @ sqlite3 
* Operating system
    * Docker: ghcr.io/muety/wakapi:2.12
* Database Sqlite

**Comment by kellpossible** [comment]
My weekly summary for projets also appears to be shifted by a day as compared to wakatime

**Comment by muety** [comment]
That's weird! Can't reproduce this, my activity chart shows every day only exactly once :thinking:. Will have a look at it, though.

**Comment by kellpossible** [comment]
I'll have another look again tomorrow! Perhaps it's something to do with my
timezone settings

**Comment by muety** [comment]
Did you perhaps change your time zone or something? Also, is either the 17th or the 18th missing instead?

**Comment by muety** [comment]
Any updates on this?

**Comment by kellpossible** [comment]
Changing timezone to UTC appears to have no effect

**Comment by kellpossible** [comment]
Here's the SVG file with the bug present, notice the tooltips for the various dates.

![wakapi-bug](https://github.com/user-attachments/assets/f0df3dfe-1e5d-4153-b384-1f03a4b4d064)

**Comment by muety** [comment]
Can you do me another favor and show the results of the following SQL query, please? 

```sql
select * from summaries where from_time between '2024-11-17' and '2024-11-20' and user_id = 'kellpossible';
```

(replace with your actual user id)

**Comment by muety** [comment]
Any updates on this?

**Comment by kellpossible** [comment]
Hi @muety sorry for the slow reply, here's the output:

```
sqlite> select * from summaries where from_time between '2024-11-17' and '2024-11-20' and user_id = 'luke';
877|luke|2024-11-17 20:05:13.021+00:00|2024-11-17 23:59:29.386+00:00|286
878|luke|2024-11-18 00:00:02.653+00:00|2024-11-18 23:57:32.284+00:00|853
879|luke|2024-11-19 00:44:03.73+00:00|2024-11-19 06:29:17.173+00:00|661
```

**Comment by muety** [comment]
Alright, found the issue! Wild story...

The root cause of all of this is the fact (TIL) that we don't store timezone information with dates. They're always stored as UTC. Of course, a zoned date is properly converted to the corresponding UTC equivalent by the driver before insertion, but nevertheless, the date's original zone information is lost. When retrieving back the date, it is, apparently, returned in whatever zone the server / application runs in (more precisely, what `TZ` environment variable is set to).

While every entity will technically still have a correct timestamp, the loss of zone information is problematic when displaying the date to an actual user. For the chart, we display the `Summary` entity's `FromTime` attribute for every day. However, for dates within the time difference between server time and user's local time, this will result in a display error.

For example:
- User is in `Europe/Berlin` (UTC+1)
- Server runs on `Europe/London` (UTC+0)
- User requests summary for `2025-01-15T00:00:00+01:00`, aka. `2025-01-14T23:00:00+00:00`
- Summary has `FromDate` set to, for example, `2025-01-14T23:30:00+00:00`
- 👉 At this point in time, it already the 15th in Berlin (as requested), but still the 14th in London
- When displaying the date to the user, they will see `2025-01-14`, even though it's - from their perspective - the summary of `2025-01-15`

As a solution, we can explicitly set every date's timezone back to the user's preference. However, I'd be surprised if this problem wouldn't affect many other parts of Wakapi as well 😐.

Will push a fix for this specific issue at hand, but will probably come back to here.

Btw. from a quick research, it looks like neither database (SQLite, MySQL, Postgres) has a data type that actually stores a date's _original_ timezone in addition.

**Comment by kellpossible** [comment]
@muety wow thanks so much for looking into this and fixing it! Sounds like it was a bit of a monster to debug in the end. I did indeed change my time zone at some point when I realised it didn't match my local one. Will it be possible for me to continue to change time zones? (I sometimes travel to new locations with different timezones and work there) I bought you a couple of coffees to say thanks!

Forgive my lack of understanding of the internals of wakapi, I had some questions:

> While every entity will technically still have a correct timestamp, the loss of zone information is problematic when displaying the date to an actual user. For the chart, we display the Summary entity's FromTime attribute for every day. However, for dates within the time difference between server time and user's local time, this will result in a display error.

Do the wakatime plugins submit times in UTC time zone? Or local times? Why not keep everything in UTC and convert just before displaying to the user, either on the frontend or some final processing step in templates?

**Comment by muety** [comment]
> Will it be possible for me to continue to change time zones?

Yes, with this new fix, you should be good to go with changing your preferred time zone as you like.

Regarding your understanding question: WakaTime plugins send raw UNIX timestamps, i.e. without any time zone information. That's also what we end up storing in the database for the most parts. Time zones _only_ come into play for display and visualization, especially in cases where a user requests intervals like _"Yesterday"_, _"Today"_, etc., which are subjective, depending on where you're at in the world. 

For generating the activity chart, for instance, we (1) retrieve the requesting user's configured tz preference, (2) resolve intervals (e.g. _"today_") accordingly, (3) fetch relevant heartbeats (with "unzoned" timestamps) from the database and (4) enrich them with the user's tz for proper display. My initial though was that, if the dates stored in the database kept their original time zone, we could get rid of the last step. But also only if the user didn't change their zone, which isn't a valid assumption (see your case).

So tl;dr: storing timestamps in UTC (or any other _fixed_ zone) consistently and retrieving and "formatting" them according to user preference is the way to go and is also what we settled for now (and also what you had suggested).
